### PR TITLE
Priority in completion

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
It turns out FCS returns completion items in interesting order in some cases, but Roslyn ignore it and sort everything alphabetically. However, it provides a way to sort as you wish. This PR pass `CompletionPriority` property for each item, which is used to sort items. 

Currently it shows arguments and writable properties names at the top of completion list:

image

Thought whether it's a good behavior or not?

### CODE REVIEW

**Comment by smoothdeveloper** [comment]
Need to think more about it, but locals (in order of proximity, closest first) should have priority IMHO, then parameters, then members (can we list static members prefixed with the type as it is required in F#?) then all the rest (like `AbstractClassAttribute`).

R# seems to list locals first (well, after null):

image

It also has that ctrl+shift+alt mode which will trim the list to expressions that match the expected type at insertion point:

image

**Comment by smoothdeveloper** [comment]
Can we get the type of expressions to display in the list? this information is really helpful.

I also wish the list was resizeable at times (when you complete on a type which has zillions of members).

**Comment by vasily-kirichenko** [comment]
We cannot adjust the list (size, show type, etc.)

**Comment by cartermp** [comment]
@vasily-kirichenko  I think this is a good change of behavior.

**Comment by vasily-kirichenko** [comment]
Own static members are shown first:

image

Own instance members are shown first:

image

Own enum fields are shown first:

image

**Comment by vasily-kirichenko** [comment]
Priorities like in R#:

* own fields
* inherited fields
* own props
* inherited props
* own methods
* inherited methods
* own events 
* inherited events
* everything else

image

**Comment by smoothdeveloper** [comment]
@vasily-kirichenko yes, correction: props come first before fields, last is postfix completion templates (we should consider that, for match, for in and some others, but not part of this PR).

Also, each time you commit a completion, it puts the item first in the list (on top of the order you listed), in decreasing order of number of times that item was commited, this is very handy.

File: vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs [hunk 1]
```diff
@@ -54,6 +54,8 @@ type internal FSharpCompletionProvider
                 | ClassificationTypeNames.NumericLiteral -> false
                 | _ -> true // anything else is a valid classification type
             ))
+
+    static let mruItems = Dictionary<string, int>()
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs]: Does this ever get cleared?

File: vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs [hunk 2]
```diff
@@ -54,6 +55,28 @@ type internal FSharpCompletionProvider
                 | ClassificationTypeNames.NumericLiteral -> false
                 | _ -> true // anything else is a valid classification type
             ))
+
+    static let mruItems = Dictionary<(* Item.FullName *) string, (* hints *) int>()
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs]: Is this ever cleared?

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs]: ?

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs]: ah, got it. No. It'll be small even if you work in VS without restart for days.

**Comment by vasily-kirichenko** [comment]
It's ready.

**Comment by vasily-kirichenko** [comment]
I've merged this branch to https://github.com/Microsoft/visualfsharp/pull/2532 because these two touch same areas of the codebase and have a lot of merge conflicts, there's no reason to keep them both.

**Pull request closed without merging**